### DISCUSSION TITLE
issue when using with Cloudflare cache

### DISCUSSION THREAD

**Discussion by gaetan-hexadog** [livewire/livewire#3112] [7. Bugs (archived)]
### Description
I'm facing the following issue `Uncaught (in promise) TypeError: initialData is null` when Cloudflare is used.
Everything works fine when bypassing (or deactivating Cloudflare cache) but the issue is raised every time (even when just updating a form field) when cache is activated.

Every component is correctly wrap with a single div.
I tried setting `key` manually on each sub-component (with unique id value).

Is there a way to get it work when Cloudflare cache is on ?

**Context**
- Livewire version: 2.3.8
- Laravel version: 8.24.0
- Alpine version: 2.8.0
- Browser: Firefox

**Comment by joshhanley** [comment]
@gaetan-hexadog I've not used Cloudflare cache. Can you tell me what it caches? I'd imaging you would want it caching any assets (JS/CSS/Images) but you wouldn't want it to cache any routes that contain a Livewire component, as it needs the server connection.

**Comment by gaetan-hexadog** [comment]
Hi,

Cloudflare cache is active automatically once website uses Cloudflare cdn purpose. Here are some details about their help page:
```
You can set Cloudflare’s CDN to cache static content according to these levels:

    No Query String: Only delivers files from cache when there is no query string.
    Ignore Query String: Delivers the same resource to everyone independent of the query string.
    Standard: Delivers a different resource each time the query string changes.

We recommend using the Standard level. Learn more

Note: By default, Cloudflare does not cache HTML content.
```

I ran some tests and there is no differences (in my use-case) between the 3 modes: issue remains present wathever the cache mode is.

So why does a cache functionnality breaks Livewire behavior ?

**Comment by stancl** [comment]
Likely that the user gets a stale version of the component props, since the fingerprint & hash are only valid for some amount of time.

You should avoid using cache on pages with LW

**Comment by gaetan-hexadog** [comment]
too bad to not being able to optimize our site performances if it's using livewire (mainly for SEO purpose). It limits the use of livewire.

**Comment by stancl** [comment]
Why do you need Livewire on SEO sensitive pages?

Also, I doubt the actual SEO benefit here. Not sure what you're doing, but Laravel in general doesn't work with CloudFlare unless you make some big config changes.

**Comment by gaetan-hexadog** [comment]
where did you see Laravel is not working with cloudflare ?

I have sites working on Laravel with cloudflare and it's working perfectly well. (and we did not had to make big config modification for it....none to be exact)

One of those site is the one with the livewire problem mentionned in the issue when I integrate livewire components (like contact form, shopping cart, shopping catalog, search, ....). So since livewire is used on this site I had to deactivate the use of cache.....

If livewire is limited to work only on backend and/or without cache system in place it should be clearly indicated into the documentation.

**Comment by stancl** [comment]
Laravel sets cookies and CloudFlare doesn't cache any responses that have cookies.

**Comment by stancl** [comment]
> If livewire is limited to work only on backend and/or without cache system in place it should be clearly indicated into the documentation.

chill

**Comment by gaetan-hexadog** [comment]
> > If livewire is limited to work only on backend and/or without cache system in place it should be clearly indicated into the documentation.
> 
> chill

didn't mean to offense you (sorry english is not my native language). I'm just saying that cache is an important thing nowadays, so if Livewire requires to not use any cache system (and that's what I understand from your answers) developper should be informed about that.

**Comment by stancl** [comment]
> cache is an important thing nowadays

Is it even working in your case? Or did you just set up CF, saw that the site didn't break (which it did with LW), and thought that it's cached?

I can't see how CF would work with Laravel without modifying the cookie headers.

**Comment by gaetan-hexadog** [comment]
> > cache is an important thing nowadays
> 
> Is it even working in your case? Or did you just set up CF, saw that the site didn't break (which it did with LW), and thought that it's cached?
> 
> I can't see how CF would work with Laravel without modifying the cookie headers.

Site was running several months with cloudflare activated and tests ran with SEO tools (google pagespeed, GTmetrix,...) showed a real difference in term of performances when clouflare was activated compared to without it.

(not sure cloudflare is the real main point of this issue because, based on what I read and understand, cloudflare is caching only static assets lile css and js).

I'm open to use another system than Clouflare but are we sure Livewire is compatible with a cache system ? or should I completely avoid using cache with Livewire ?

have you seen site using livewire working well with CDN and cache (any other system than cloudflare) ?

**Comment by stancl** [comment]
> cloudflare is caching only static assets lile css and js

In that case it's not caching your responses, so it's a completely different thing.

If LW doesn't work, but everything else (including POST forms with `@csrf`) does, then the issue is likely that LW scripts don't get loaded correctly. Check the browser console.

**Comment by gaetan-hexadog** [comment]
I'll try to find a way to make a quick test with a simpler cdn than the full clouflare solution to be sure this part is working well. if yes then I'll avoid using Cloudflare.

**Comment by gaetan-hexadog** [comment]
> > cloudflare is caching only static assets lile css and js
> 
> In that case it's not caching your responses, so it's a completely different thing.
> 
> If LW doesn't work, but everything else (including POST forms with `@csrf`) does, then the issue is likely that LW scripts don't get loaded correctly. Check the browser console.

yes that the current behavior. I'll check that.

**Comment by joshhanley** [comment]
I agree with @stancl, caching a dynamic page doesn't make sense, as there is no way for CF to know if any data has changed in the backend (ie new data in database) and when to respond with a cached version vs getting a new one from the server. I'd be surprised if Laravel pages were cached for the same reason. 

So I don't think there needs to be anything in the docs, any more than it should be in the Laravel docs, due to the reasons above. 

But I do agree with using CF to cache assets, in which case, publishing Livewire's assets and pushing them to CF would certainly help improve performance.

If you can get a sample app working with CF that demonstrates the issue, and listing what your CF is configured to cache here, then someone can hit that endpoint have a look to see what might be causing the issue.

For what it's worth, my website/blog is running Livewire without CF or any caching and it's gets pretty good results. So with a targeted asset caching strategy it should run great.

Hope this helps!

![image](https://user-images.githubusercontent.com/882837/108258936-a7e84080-71b4-11eb-9b50-4abd2deba315.png)

**Comment by gaetan-hexadog** [comment]
thanks @joshhanley for your feedback. I totally agree with both of you regarding the non sense of caching dynamic page (I have explianed myself in the wrong way at first).  My issue was with static assets cached by CF leading to LW issue. I'll try a CF setup with a very basic example to see if the issue still occurs and if so I'll share it here. thanks you both for your time and answers.

**Comment by stancl** [comment]
I'm not sure what the issue is though, you don't even need CF to cache Livewire assets. They get cached by default.

On my latest site I'm not using any caching services, and the Livewire assets are stored in browser memory.

Livewire sets a cache header in the source code:

https://github.com/livewire/livewire/blob/c0c78f430e4151fb6902bbb88c75e782d2549e1a/src/Controllers/CanPretendToBeAFile.php#L11

To me it seems like you're blindly adding services to your app without actually checking what benefit they provide. CF won't cache server responses, and everything else that should be cached (like Livewire assets) — is cached anyway.

**Comment by joshhanley** [comment]
@gaetan-hexadog no worries 🙂  if you do manage to get it working, can you also report that back here and what you did to get it working, so others can benefit from it in the future?

**Comment by gaetan-hexadog** [comment]
> @gaetan-hexadog no worries 🙂  if you do manage to get it working, can you also report that back here and what you did to get it working, so others can benefit from it in the future?

yes sure.

**Comment by gaetan-hexadog** [comment]
Unfortunately I did not succeed in reproducing the case. When I create a new project with a component that tries to reproduce the same behavior it works perfectly, but with the complete application it doesn't. It must be due to an edge effect related to the complexity of my application and the component encapsulations.

**Comment by helturkey** [comment]
I have disabled cloudflare html minification and it works fine, disabled google pagespeed html remove comments too.

**Comment by ChaerilM** [comment]
help, i am having problem with this as well, the website getting 419 from livewire/update but refreshing does not do anything. 
i don't have minify, cache level standard, and my cloudflare cache is set to cache whole hostname, except uri contains livewire. but its still doesn't work.

**Reply by lrljoe** [reply]
419 will be related to your CSRF field, or perhaps your session domain is incorrectly set?

I have plenty of instances behind CloudFlare and other WAFs, without issue.,

**Comment by CodyPChristian** [comment]
To anyone who comes across this like I did. Like @stancl and @joshhanley said its best to not use CF with Laravel or Livewire but its not impossible either. You can also opt to set a custom rule in CF to only cache your static assets and leave actual pages alone using the bypass cache option under the custom rule. Additionally if you need 'cache' then just use Spaties responsecache package on your views and you'll be set. We heavily use CF with all of our Laravel/Livewire apps and simply just bypass the cache on anything non-static while still able to use the rest of CFs features like WAF, SSL, Geo Headers, etc.
